* [Trinity Wiki Home](Home)
* [Installing Trinity](Installing-Trinity)
    * [Trinity Computing Requirements](Trinity-Computing-Requirements)
    * [Accessing Trinity on Publicly Available Compute Resources](Accessing-Trinity-on-Publicly-Available-Compute-Resources)
    * [Run Trinity using Docker](Trinity-in-Docker)
    * [Run Trinity using Singularity](Trinity-in-Docker#trinity_singularity)
* [Running Trinity](Running-Trinity)
    * [Genome Guided Trinity Transcriptome Assembly](Genome-Guided-Trinity-Transcriptome-Assembly)
    * [Gene Structure Annotation of Genomes](Genome-Structure-Annotation-Using-Trinity--and-PASA)
* Trinity process and resource monitoring
    * [Monitoring Progress During a Trinity Run](Trinity-Progress-Monitoring)
    * [Examining Resource Usage at the End of a Trinity Run](Trinity-Runtime-Profiling)
* [Output of Trinity Assembly](Output-of-Trinity-Assembly)
* [Assembly Quality Assessment](Transcriptome-Assembly-Quality-Assessment)
    * [Counting Full-length Transcripts](Counting-Full-Length-Trinity-Transcripts)
    * [RNA-Seq Read Representation](RNA-Seq-Read-Representation-by-Trinity-Assembly)
    * [Contig Nx and ExN50 stats](Transcriptome-Contig-Nx-and-ExN50-stats)
    * [Examine strand-specificity of reads](Examine-Strand-Specificity)
* [Downstream Analyses](Post-Transcriptome-Assembly-Downstream-Analyses)
    * [Transcript Quantification](Trinity-Transcript-Quantification)
    * [QC Samples and Bio Replicates](QC-Samples-and-Biological-Replicates)
    * [Differential Transcript or Gene Expression](Trinity-Differential-Expression)
        * [Sample Specificity Analysis in Many Sample Comparisons](Sample-Specific-Expression)
    * [SuperTranscripts](SuperTranscripts)
         * [Differential Transcript Usage](DiffTranscriptUsage)
         * [Identifying Sequence Polymorphisms or Variants](Variant-Calling)
    * [Coding Region Identification](Coding-Region-Identification-in-Trinity-Assemblies)
    * [Functional Annotation of Transcripts](Functional-Annotation-of-Transcripts)
    * [Gene Ontology term functional category enrichments](Running-GOSeq)
* Miscellaneous additional functionality that may be of interest
    * [In silico normalization of fastq files](Trinity-Insilico-Normalization)
* [Contributing code](Contributing)
    * [Contributing guides and protocols](https://github.com/trinityrnaseq/trinity_community_codebase/wiki)
        * [Defining a reduced 'best' transcript set and TSA submission](https://github.com/trinityrnaseq/trinity_community_codebase/wiki/Trinity-best-transcript-set)
* [Trinity Tidbits](Trinity-Tidbits)
* Frequently Asked Questions (FAQ)
    * [There are too many transcripts! What do I do?](https://github.com/trinityrnaseq/trinityrnaseq/wiki/There-are-too-many-transcripts!-What-do-I-do%3F)
    * [How do I use reads I downloaded from SRA](https://github.com/trinityrnaseq/trinityrnaseq/wiki/How-do-I-use-reads-I-downloaded-from-SRA%3F)
    * [How do I identify the specific reads that were incorporated into the transcript assemblies?](https://github.com/trinityrnaseq/trinityrnaseq/wiki/How-do-I-identify-the-specific-reads-that-were-incorporated-into-the-transcript-assemblies%3F)
    * [How do I combine reads?](https://github.com/trinityrnaseq/trinityrnaseq/wiki/How-do-I-combine-reads%3F)
    * [How can I run this in parallel on a computing grid?](https://github.com/trinityrnaseq/trinityrnaseq/wiki/How-can-I-run-this-in-parallel-on-a-computing-grid%3F)
    * [Computing and Time requirements](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Computing-and-Time-requirements)
    * [Errors during Trinity run](https://github.com/trinityrnaseq/trinityrnaseq/wiki/Errors-during-Trinity-run)